# **Predicting Taxi Gratuities in NYC**
**Stage 1 - Preliminary data analysis**

**Situatiom:**
We're still in the early stages of the project. New York City TLC has given us access to their data. To get clear insights, New York TLC's data must be inspected, key variables identified, and the dataset prepared for analysis.

# Inspecting and analyzing data

In this stage, we will examine the provided data and prepare it for analysis. This activity will help us ensure the information is:

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>    

**The purpose** of this stage is to investigate and understand the dataset.
  
**The goal** is to use a dataframe constructed within Python, perform a cursory inspection of the dataset, and inform team members of the findings.
<br/>  
*This stage has three parts:*

**Part 1:** Understanding the situation
* Prepare to understand and organize the taxi cab dataset and information.

**Part 2:** Understanding the data

* Creating a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compiling summary information about the data to inform next steps.

**Part 3:** Understanding variables

* Using insights from the examination of the summary data to guide deeper investigation into specific variables.

# **Identifying data types and relevant variables using Python**


# **PACE stages**


## PACE: **Plan**

### **Part 1. Understanding the situation**

> By considering the overall goals of the project, reviewing the project proposal and the necessary syntax for inspecting and preparing the data, and importing the data to take an overview, we can best prepare to understand and organize the taxi cab information.

## PACE: **Analyze**

### **Part 2a. Building the dataframe**

In [ ]:
# Importing libraries and packages listed above
import numpy as np
import pandas as pd

# Loading dataset into dataframe
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
print("done")

done


### **Part 2b. Understanding the data - Inspecting the data**

1. We noticed that there aren't any null values in the dataset, only 3 columns are of type object, and the rest of the 15 columns are all numeric values. Also, the first column doesn't have a name.
2. The columns 'fare_amount', 'tip_amount', and 'total_amount' have outliers.

In [ ]:
df.head(10)

,Unnamed: 0,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,24870114,2,03/25/2017 8:55:43 AM,03/25/2017 9:09:47 AM,6,3.34,1,N,100,231,1,13.0,0.0,0.5,2.76,0.0,0.3,16.56
1,35634249,1,04/11/2017 2:53:28 PM,04/11/2017 3:19:58 PM,1,1.80,1,N,186,43,1,16.0,0.0,0.5,4.00,0.0,0.3,20.80
2,106203690,1,12/15/2017 7:26:56 AM,12/15/2017 7:34:08 AM,1,1.00,1,N,262,236,1,6.5,0.0,0.5,1.45,0.0,0.3,8.75
3,38942136,2,05/07/2017 1:17:59 PM,05/07/2017 1:48:14 PM,1,3.70,1,N,188,97,1,20.5,0.0,0.5,6.39,0.0,0.3,27.69
4,30841670,2,04/15/2017 11:32:20 PM,04/15/2017 11:49:03 PM,1,4.37,1,N,4,112,2,16.5,0.5,0.5,0.00,0.0,0.3,17.80
5,23345809,2,03/25/2017 8:34:11 PM,03/25/2017 8:42:11 PM,6,2.30,1,N,161,236,1,9.0,0.5,0.5,2.06,0.0,0.3,12.36
6,37660487,2,05/03/2017 7:04:09 PM,05/03/2017 8:03:47 PM,1,12.83,1,N,79,241,1,47.5,1.0,0.5,9.86,0.0,0.3,59.16
7,69059411,2,08/15/2017 5:41:06 PM,08/15/2017 6:03:05 PM,1,2.98,1,N,237,114,1,16.0,1.0,0.5,1.78,0.0,0.3,19.58
8,8433159,2,02/04/2017 4:17:07 PM,02/04/2017 4:29:14 PM,1,1.20,1,N,234,249,2,9.0,0.0,0.5,0.00,0.0,0.3,9.80
9,95294817,1,11/10/2017 3:20:29 PM,11/10/2017 3:40:55 PM,1,1.60,1,N,239,237,1,13.0,0.0,0.5,2.75,0.0,0.3,16.55


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [ ]:
stats = df.describe()
stats.loc['interval']= stats.loc['max'] - stats.loc['min']

### **Part 2c. Understanding the data - Investigating the variables**

We'll sort and interpret the data table for two key variables: `trip_distance` and `total_amount`.

1. Sorting the first variable (`trip_distance`) from maximum to minimum value, we found that the values are greatly larger than the mean.
2. Sorting by the second variable (`total_amount`), we observed outliers and negative values.
3. Greater distance doesn't seem to reflect a higher total amount in the results.

In [ ]:
# Sorting the data by trip distance from maximum to minimum value
df.sort_values('trip_distance', ascending=False).head()

,Unnamed: 0,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
9280,51810714,2,06/18/2017 11:33:25 PM,06/19/2017 12:12:38 AM,2,33.96,5,N,132,265,2,150.00,0.0,0.0,0.00,0.00,0.3,150.30
13861,40523668,2,05/19/2017 8:20:21 AM,05/19/2017 9:20:30 AM,1,33.92,5,N,229,265,1,200.01,0.0,0.5,51.64,5.76,0.3,258.21
6064,49894023,2,06/13/2017 12:30:22 PM,06/13/2017 1:37:51 PM,1,32.72,3,N,138,1,1,107.00,0.0,0.0,55.50,16.26,0.3,179.06
10291,76319330,2,09/11/2017 11:41:04 AM,09/11/2017 12:18:58 PM,1,31.95,4,N,138,265,2,131.00,0.0,0.5,0.00,0.00,0.3,131.80
29,94052446,2,11/06/2017 8:30:50 PM,11/07/2017 12:00:00 AM,1,30.83,1,N,132,23,1,80.00,0.5,0.5,18.56,11.52,0.3,111.38


In [ ]:
# Sorting the data by total amount and printing the top 20 values
df.sort_values('total_amount', ascending=False).head(20)

,Unnamed: 0,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
8476,11157412,1,02/06/2017 5:50:10 AM,02/06/2017 5:51:08 AM,1,2.60,5,N,226,226,1,999.99,0.0,0.0,200.00,0.00,0.3,1200.29
20312,107558404,2,12/19/2017 9:40:46 AM,12/19/2017 9:40:55 AM,2,0.00,5,N,265,265,2,450.00,0.0,0.0,0.00,0.00,0.3,450.30
13861,40523668,2,05/19/2017 8:20:21 AM,05/19/2017 9:20:30 AM,1,33.92,5,N,229,265,1,200.01,0.0,0.5,51.64,5.76,0.3,258.21
15474,55538852,2,06/06/2017 8:55:01 PM,06/06/2017 8:55:06 PM,1,0.00,5,N,265,265,1,200.00,0.0,0.5,11.00,0.00,0.3,211.80
12511,107108848,2,12/17/2017 6:24:24 PM,12/17/2017 6:24:42 PM,1,0.00,5,N,265,265,1,175.00,0.0,0.0,46.69,11.75,0.3,233.74
3582,111653084,1,01/01/2017 11:53:01 PM,01/01/2017 11:53:42 PM,1,7.30,5,N,1,1,1,152.00,0.0,0.0,0.00,0.00,0.3,152.30
9280,51810714,2,06/18/2017 11:33:25 PM,06/19/2017 12:12:38 AM,2,33.96,5,N,132,265,2,150.00,0.0,0.0,0.00,0.00,0.3,150.30
16379,101198443,2,11/30/2017 10:41:11 AM,11/30/2017 11:31:45 AM,1,25.50,5,N,132,265,2,140.00,0.0,0.5,0.00,16.26,0.3,157.06
10291,76319330,2,09/11/2017 11:41:04 AM,09/11/2017 12:18:58 PM,1,31.95,4,N,138,265,2,131.00,0.0,0.5,0.00,0.00,0.3,131.80
11269,51920669,1,06/19/2017 12:51:17 AM,06/19/2017 12:52:12 AM,2,0.00,5,N,265,265,1,120.00,0.0,0.0,20.00,11.52,0.3,151.82


In [ ]:
# Sorting the data by total amount and printing the bottom 20 values
df.sort_values('total_amount').head(20)
df[df['total_amount'] < 0].groupby('payment_type').size()

payment_type
3    7
4    7
dtype: int64

In [ ]:
# Checking how many of each payment type are represented in the data
df.groupby('payment_type').size()

payment_type
1    15265
2     7267
3      121
4       46
dtype: int64

According to the data dictionary, the payment method was encoded as follows:

1 = Credit card  
2 = Cash  
3 = No charge  
4 = Dispute  
5 = Unknown  
6 = Voided trip

In [ ]:
# Finding the average tip for trips paid for with a credit card
print(df[df['payment_type'] == 1]['tip_amount'].mean())

# Finding the average tip for trips paid for with cash
df[df['payment_type'] == 2]['tip_amount'].mean()

2.7298001965279934


0.0

In [ ]:
# Checking how many times each vendor ID is represented in the data
df.groupby('VendorID').size()

VendorID
1    10073
2    12626
dtype: int64

In [ ]:
# Calculating the mean total amount for each vendor
df.groupby('VendorID')['total_amount'].mean()

VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64

In [ ]:
# Filtering the data for credit card payments only
cc_df = df[df['payment_type'] == 1]

# Filtering credit-card-only data for passenger count only
pcc_df = cc_df[['passenger_count']]

,passenger_count
0,6
1,1
2,1
3,1
5,6
...,...
22692,1
22693,1
22695,1
22697,1


In [ ]:
# Calculating the average tip amount for each passenger count (credit card payments only)
cc_df.groupby('passenger_count')['tip_amount'].mean()

passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64

## PACE: **Construct**
We have determined that the Construct stage does not apply to this specific stage.

## PACE: **Execute**



### **Given these efforts, what can we summarize for the data team?**

+ We identified the following data types in the dataset:
   - Eight columns of type float64
   - Seven columns of type int64
   - Three columns of type object
+ We verified that the data contains 18 columns and 22,699 rows with no missing or null values.
+ We identified that the 'fare_amount', 'tip_amount', and 'total_amount' columns have outliers.
+ We observed that greater trip distance doesn't seem to reflect a higher total amount, showing a lack of direct correlation between these two parameters in some cases.
+ We found that most users paid using credit cards, and the average tip amount for cash payments in the dataset is zero.
+ We found no strong correlation between the number of passengers and the tip amount.
+ We spotted 14 records with negative total amounts, which were caused by either no charge or dispute methods.
+ We concluded that the two most important variables for the predictive model are total amount and trip distance.